zkevm-circuits/src/mpt_circuit/storage_leaf/leaf_key.rs

```diff
@@ -86,873 +90,120 @@ kind of case we have:
+            }};
+
+            // Calculate and store the key RLC.
+            let key_rlc = key_rlc_prev.expr()
+                + storage.key_rlc(
                     meta,
-                    s_main.bytes.to_vec(),
-                    1,
-                    one.clone(),
-                    0,
-                    power_of_randomness.clone(),
+                    &mut cb.base,
+                    key_mult_prev,
+                    is_key_odd.expr(),
+                    1.expr(),
+                    true,
                 );
-
-            let c_rlp1 = meta.query_advice(c_main.rlp1, Rotation::cur());
-            // c_rlp2 can appear if long and if no branch above leaf
-            let c_rlp2 = meta.query_advice(c_main.rlp2, Rotation::cur());
-            rlc = rlc
-                + c_rlp1
-                    * power_of_randomness[POWER_OF_RANDOMNESS_LEN - 1].clone()
-                    * power_of_randomness[1].clone();
-            rlc = rlc
-                + c_rlp2
-                    * power_of_randomness[POWER_OF_RANDOMNESS_LEN - 1].clone()
-                    * power_of_randomness[2].clone();
-
-            let acc = meta.query_advice(accs.acc_s.rlc, Rotation::cur());
-
-            /*
-            We need to ensure that the RLC of the row is computed properly for `is_short` and
-            `is_long`. We compare the computed value with the value stored in `accumulators.acc_s.rlc`.
-            */
-            constraints.push((
-                "Leaf key RLC (short or long)",
-                q_enable.clone()
-                    * (is_short + is_long)
-                    * (one.clone() - is_leaf_placeholder.clone())
-                    * (rlc - acc.clone()),
-            ));
-
-            /*
-            We need to ensure that the RLC of the row is computed properly for `last_level` and
-            `one_nibble`. We compare the computed value with the value stored in `accumulators.acc_s.rlc`.
-
-            `last_level` and `one_nibble` cases have one RLP byte (`s_rlp1`) and one byte (`s_rlp2`)
-            where it is 32 (for `last_level`) or `48 + last_nibble` (for `one_nibble`).
-            */
-            constraints.push((
-                "Leaf key RLC (last level or one nibble)",
-                q_enable
-                    * (last_level + one_nibble)
-                    * (one.clone() - is_leaf_placeholder)
-                    * (rlc_last_level_or_one_nibble - acc),
-            ));
-
-            constraints
+            require!(a!(accs.key.rlc) => key_rlc);
+
+            // Total number of nibbles needs to be KEY_LEN_IN_NIBBLES (except in a
+            // placeholder leaf).
+            // TODO(Brecht): why not in placeholder leaf?
+            ifx! {not!(storage.is_placeholder(meta)) => {
+                let num_nibbles = storage.num_key_nibbles(meta, is_key_odd.expr());
+                require!(nibbles_count_prev + num_nibbles => KEY_LEN_IN_NIBBLES);
+            }}
```

Comment:
Storage leaf placeholder didn't have proper RLP encoding, so this could fail. Now fixed in witness generator and the corresponding changes in circuit code made in https://github.com/Brechtpd/zkevm-circuits/pull/6/commits/0e23861ebd479cb5b62f29d01566c4c79846e203.